js/script.js

```diff
@@ -0,0 +1,117 @@
+
+GameBoard = (function () {
+    let gameBoardViewContainer = null;
+    let playerOne = null;
+    let playerTwo = null;
+    let hasGameStarted = false
+    const isWinningMove = (gameboard) => {
+        switch (true) {
+            case (gameboard[0] === 'O' && gameboard[1] === 'O' && gameboard[2] === 'O') || (gameboard[0] === 'X' && gameboard[1] === 'X' && gameboard[2] === 'X'):
+                return true
+            case (gameboard[3] === 'O' && gameboard[4] === 'O' && gameboard[5] === 'O') || (gameboard[3] === 'X' && gameboard[4] === 'X' && gameboard[5] === 'X'):
+                return true
+            case (gameboard[6] === 'O' && gameboard[7] === 'O' && gameboard[8] === 'O') || (gameboard[6] === 'X' && gameboard[7] === 'X' && gameboard[8] === 'X'):
+                return true
+            case (gameboard[0] === 'O' && gameboard[3] === 'O' && gameboard[6] === 'O') || (gameboard[0] === 'X' && gameboard[3] === 'X' && gameboard[6] === 'X'):
+                return true
+            case (gameboard[1] === 'O' && gameboard[4] === 'O' && gameboard[7] === 'O') || (gameboard[1] === 'X' && gameboard[4] === 'X' && gameboard[7] === 'X'):
+                return true
+            case (gameboard[2] === 'O' && gameboard[5] === 'O' && gameboard[8] === 'O') || (gameboard[2] === 'X' && gameboard[5] === 'X' && gameboard[8] === 'X'):
+                return true
+            case (gameboard[0] === 'O' && gameboard[4] === 'O' && gameboard[8] === 'O') || (gameboard[0] === 'X' && gameboard[4] === 'X' && gameboard[8] === 'X'):
+                return true
+            case (gameboard[2] === 'O' && gameboard[4] === 'O' && gameboard[6] === 'O') || (gameboard[2] === 'X' && gameboard[4] === 'X' && gameboard[6] === 'X'):
+                return true
+        }
+    }
+    const isTie = (gameboard) => gameboard.filter(gameSpot => gameSpot === null).length === 0
+    const reDrawOnBoard = (gameboard, position, marker) => {
+      const newGameboardDrawing = gameboard.map((item, index) => {
+        if (index === position) {
+            return marker
+        }
+        return item;
+      });
+      if (isWinningMove(newGameboardDrawing)) {
+        if (playerOne.getPlayerTurn()) {
+            alert(playerOne.getPlayerName()+ ' is the winner')
+        }
+        else if (playerTwo.getPlayerTurn()) {
+            alert(playerTwo.getPlayerName()+ ' is the winner')
+        }
+        playerOne.setPlayerTurn(false);
+        playerTwo.setPlayerTurn(false)
+        hasGameStarted = false
+      }
+      if (isTie(newGameboardDrawing)) {
+        hasGameStarted = false
+        alert('Is a tie')
+      }
+      renderGameBoard(newGameboardDrawing)
+    }
```

Comment:
This seems like a lot. You generate a new board array with the latest move and check if that move is going to win or if it's a tie. Why did you go with this implementation?

Reply:
We are trying to follow the functional programming paradigm. So we didnt want the function to mutate the data, but rather create a new data. What is why we were using filter and map function a lot.